src/Query/Builder.php

```diff
@@ -1086,28 +1090,33 @@ protected function compileWhereBasic(array $where): array
             $regex = preg_replace('#(^|[^\\\])%#', '$1.*', preg_quote($value));
 
             // Convert like to regular expression.
-            if (! Str::startsWith($value, '%')) {
+            if (! str_starts_with($value, '%')) {
                 $regex = '^'.$regex;
             }
-            if (! Str::endsWith($value, '%')) {
+            if (! str_ends_with($value, '%')) {
                 $regex .= '$';
             }
 
-            $value = new Regex($regex, 'i');
-        } // Manipulate regexp operations.
-        elseif (in_array($operator, ['regexp', 'not regexp', 'regex', 'not regex'])) {
+            $value = new Regex($regex, $flags);
+        } // Manipulate regex operations.
+        elseif (in_array($operator, ['regex', 'not regex'])) {
             // Automatically convert regular expression strings to Regex objects.
             if (! $value instanceof Regex) {
                 $e = explode('/', $value);
-                $flag = end($e);
-                $regstr = substr($value, 1, -(strlen($flag) + 1));
-                $value = new Regex($regstr, $flag);
+                if ($value[0] !== '/' || count($e) < 3) {
+                    throw new \InvalidArgumentException(sprintf('Regular expressions must be surrounded by slashes "/". Got "%s"', $value));
+                }
+                $flags = end($e);
+                $regstr = substr($value, 1, -1 - strlen($flags));
```

Comment:
I just want to confirm that this handles cases where the pattern includes escaped forward slashes. If we don't currently test that perhaps we should.

I _think_ it does, you only assert that `explode()` returns at least three components and only the final component is used to extra flags (which will be an empty string if something like `/acme/` is passed in).

I think there may be some edge cases where parsing might accept invalid input. Given `'/acme/bar/foo'`, this will construct a Regex with the pattern `acme/bar`, which seems incorrect? I'm not sure that the Laravel library should care about that, though. Probably not.

----

If it helps, here are some local tests using a BSON regex object where the pattern string contains slash characters.

```php
<?php

require __DIR__ . '/vendor/autoload.php';

$client = new MongoDB\Client('mongodb://localhost:27070/?replicaSet=rs0');
$collection = $client->test->foo;
$collection->drop();

$collection->insertOne(['x' => 'foo']);
$collection->insertOne(['x' => 'foo/']);

foreach (['foo', 'foo/', 'foo\/', 'foo\\/', 'foo\\\/'] as $pattern) {
    $regex = new MongoDB\BSON\Regex($pattern);
    $cursor = $collection->find(['x' => $regex]);
    $cursor->setTypeMap(['root' => 'bson']);

    printf("Querying with: %s\n", json_encode($regex));

    foreach ($cursor as $document) {
        echo $document->toRelaxedExtendedJSON(), "\n";
    }

    echo "\n";
}
```

Output:

```
$ php foo.php 
Querying with: {"$regex":"foo","$options":""}
{ "_id" : { "$oid" : "64b97a01285f15fae9083a62" }, "x" : "foo" }
{ "_id" : { "$oid" : "64b97a01285f15fae9083a63" }, "x" : "foo/" }

Querying with: {"$regex":"foo\/","$options":""}
{ "_id" : { "$oid" : "64b97a01285f15fae9083a63" }, "x" : "foo/" }

Querying with: {"$regex":"foo\\\/","$options":""}
{ "_id" : { "$oid" : "64b97a01285f15fae9083a63" }, "x" : "foo/" }

Querying with: {"$regex":"foo\\\/","$options":""}
{ "_id" : { "$oid" : "64b97a01285f15fae9083a63" }, "x" : "foo/" }

Querying with: {"$regex":"foo\\\\\/","$options":""}

```

Reply:
@GromNaN: I think the only possible action item from the above is:

> I just want to confirm that this handles cases where the pattern includes escaped forward slashes. If we don't currently test that perhaps we should.

I'll defer to you. If you don't think any such test is necessary, feel free to resolve this thread after reading.

Reply:
I believe any more regex parsing than absolutely necessary should be avoided. Looking at the example with `/acme/bar/foo`, plugging that into `preg_match` reveals that the first unescaped forward slash is treated as the end delimiter, resulting in `b` being reported as an unknown modified: https://3v4l.org/sk2fQ

Anything more than that would require us to specifically explode on unescaped delimiters. This means that simply matching for `'[^\\\]' . $delimiter` won't work, as the backslash itself could've been escaped, resulting once again in an unescaped delimiter. There's only so much polishing one can do on crappy inputs to make them work.

Reply:
> plugging that into preg_match reveals that the first unescaped forward slash is treated as the end delimiter

Would it make sense to use consistent parsing here? So we pick up the first delimiter character and then find the next unescaped occurrence? It's unfortunate there's no `preg_unquote()` implementation, but perhaps `stripslashes()` might suffice? (see: https://stackoverflow.com/a/9172908/162228).

If we rely on that, we're more likely to produce an exception by yielding an invalid flag string, instead of a pattern that won't match anything and could be more confusing to debug.

Reply:
Quick follow-up as I mentioned in Slack while thinking about this some more:

> Although to be honest I'm not confident `stripslashes()` would be a fool-proof solution, as `stripslashes('/foo\/bar/i')` yields `"/foo/bar/i"` and it's still ambiguous what the user intended because we eliminated the escaping. So maybe this problem is closer to the parsing you had to do to deal with `%` and `_` escaping for like.

Given that, I now understand @alcaeus' previous comment:

> Anything more than that would require us to specifically explode on unescaped delimiters.

I think we're OK to leave this as-is, but I'd propose adding a comment somewhere around this code to note the edge cases we're choosing _not_ to handle.